File: parqueadero_app/lib/features/turnos/domain/turno.dart
enum EstadoTurno {
  abierto,
  cerradoPendienteArqueo,
  cerrado;

  static EstadoTurno fromBackend(String value) => switch (value) {
    'ABIERTO' => EstadoTurno.abierto,
    'CERRADO_PENDIENTE_ARQUEO' => EstadoTurno.cerradoPendienteArqueo,
    'CERRADO' => EstadoTurno.cerrado,
    _ => throw FormatException('estado de turno desconocido recibido del backend: $value'),
  };

  String toBackend() => switch (this) {
    EstadoTurno.abierto => 'ABIERTO',
    EstadoTurno.cerradoPendienteArqueo => 'CERRADO_PENDIENTE_ARQUEO',
    EstadoTurno.cerrado => 'CERRADO',
  };
}

class Turno {
  const Turno({
    required this.id,
    required this.operadorId,
    required this.apertura,
    this.cierre,
    required this.baseInicial,
    this.totalRecaudado,
    this.efectivoContado,
    this.efectivoEsperado,
    this.diferencia,
    required this.estado,
    this.validadoPorId,
    this.validadoEn,
    required this.createdAt,
    required this.updatedAt,
  });

  final String id;
  final String operadorId;
  final DateTime apertura;
  final DateTime? cierre;
  final int baseInicial;
  final int? totalRecaudado;
  /// Los siguientes tres solo se llenan al cerrar el turno.
  final int? efectivoContado;
  final int? efectivoEsperado;
  final int? diferencia;
  final EstadoTurno estado;

  /// Solo distintos de `null` si el turno pasó por
  /// [EstadoTurno.cerradoPendienteArqueo] y un ADMIN ya completó el arqueo
  /// (mismo criterio que en `ArqueoTurno`).
  final String? validadoPorId;
  final DateTime? validadoEn;
  final DateTime createdAt;
  final DateTime updatedAt;
}

File: parqueadero-api/src/models/turno.js
export class Turno {
  constructor({
    id,
    operadorId,
    apertura,
    cierre,
    baseInicial,
    totalRecaudado,
    efectivoContado,
    efectivoEsperado,
    diferencia,
    estado,
    validadoPorId,
    validadoEn,
    createdAt,
    updatedAt,
  }) {
    this.id = id;
    this.operadorId = operadorId;
    this.apertura = apertura;
    this.cierre = cierre;
    this.baseInicial = baseInicial;
    this.totalRecaudado = totalRecaudado;
    this.efectivoContado = efectivoContado;
    this.efectivoEsperado = efectivoEsperado;
    this.diferencia = diferencia;
    this.estado = estado;
    this.validadoPorId = validadoPorId;
    this.validadoEn = validadoEn;
    this.createdAt = createdAt;
    this.updatedAt = updatedAt;
  }

  static toDomain(record) {
    return new Turno({
      id: record.id,
      operadorId: record.operadorId,
      apertura: record.apertura,
      cierre: record.cierre,
      baseInicial: record.baseInicial,
      totalRecaudado: record.totalRecaudado,
      efectivoContado: record.efectivoContado,
      efectivoEsperado: record.efectivoEsperado,
      diferencia: record.diferencia,
      estado: record.estado,
      validadoPorId: record.validadoPorId,
      validadoEn: record.validadoEn,
      createdAt: record.createdAt,
      updatedAt: record.updatedAt,
    });
  }

  static toPersistence({
    operadorId,
    baseInicial,
    cierre,
    totalRecaudado,
    efectivoContado,
    efectivoEsperado,
    diferencia,
    estado,
    validadoPorId,
    validadoEn,
  } = {}) {
    const data = {};
    if (operadorId !== undefined) data.operadorId = operadorId;
    if (baseInicial !== undefined) data.baseInicial = baseInicial;
    if (cierre !== undefined) data.cierre = cierre;
    if (totalRecaudado !== undefined) data.totalRecaudado = totalRecaudado;
    if (efectivoContado !== undefined) data.efectivoContado = efectivoContado;
    if (efectivoEsperado !== undefined) data.efectivoEsperado = efectivoEsperado;
    if (diferencia !== undefined) data.diferencia = diferencia;
    if (estado !== undefined) data.estado = estado;
    if (validadoPorId !== undefined) data.validadoPorId = validadoPorId;
    if (validadoEn !== undefined) data.validadoEn = validadoEn;
    return data;
  }
}
